federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java

```diff
@@ -434,6 +428,64 @@ public List<UserModel> loadUsersByUsernames(List<String> usernames, RealmModel r
         return result;
     }
 
+    /**
+     * Searches LDAP using logical conjunction of params. It supports 
+     * <ul>
+     *     <li>{@link UserModel#FIRST_NAME}</li>
+     *     <li>{@link UserModel#LAST_NAME}</li>
+     *     <li>{@link UserModel#EMAIL}</li>
+     *     <li>{@link UserModel#USERNAME}</li>
+     * </ul>
+     * 
+     * For zero or any other param it returns all users.
+     */
+    private List<LDAPObject> searchLDAPByAttributes(RealmModel realm, Map<String, String> attributes) {
+
+        if (attributes.containsKey(UserModel.USERNAME) || 
+            attributes.containsKey(UserModel.EMAIL) || 
+            attributes.containsKey(UserModel.FIRST_NAME) || 
+            attributes.containsKey(UserModel.LAST_NAME)) {
+
+            try (LDAPQuery ldapQuery = LDAPUtils.createQueryForUserSearch(this, realm)) {
+                LDAPQueryConditionsBuilder conditionsBuilder = new LDAPQueryConditionsBuilder();
+
+                // Mapper should replace parameter with correct LDAP mapped attributes
+                if (attributes.containsKey(UserModel.USERNAME)) {
+                    ldapQuery.addWhereCondition(conditionsBuilder.equal(UserModel.USERNAME, attributes.get(UserModel.USERNAME), EscapeStrategy.NON_ASCII_CHARS_ONLY));
+                }
+                if (attributes.containsKey(UserModel.EMAIL)) {
+                    ldapQuery.addWhereCondition(conditionsBuilder.equal(UserModel.EMAIL, attributes.get(UserModel.EMAIL), EscapeStrategy.NON_ASCII_CHARS_ONLY));
+                }
+                if (attributes.containsKey(UserModel.FIRST_NAME)) {
+                    ldapQuery.addWhereCondition(conditionsBuilder.equal(UserModel.FIRST_NAME, attributes.get(UserModel.FIRST_NAME), EscapeStrategy.NON_ASCII_CHARS_ONLY));
+                }
+                if (attributes.containsKey(UserModel.LAST_NAME)) {
+                    ldapQuery.addWhereCondition(conditionsBuilder.equal(UserModel.LAST_NAME, attributes.get(UserModel.LAST_NAME), EscapeStrategy.NON_ASCII_CHARS_ONLY));
+                }
+
+                return ldapQuery.getResultList();
+            }
+
+        } else {
+            // for all other searchable fields: Ignoring is the fallback option, since it may overestimate the results but does not ignore matches.
+            // for empty params: all users are returned
+            try (LDAPQuery ldapQuery = LDAPUtils.createQueryForUserSearch(this, realm)) {
+                return ldapQuery.getResultList();
```

Comment:
AFAIK When you search in the admin console and you display the users (in the old admin console, when you clicked on the button "View all users" in the old admin console):
- In case of LDAP import mode, it shows only LDAP users, which were already imported into Keycloak DB before (usually users which login at least once into Keycloak etc)
- In case of LDAP no-import mode, it doesn't show any LDAP users at all

You need to add something into the "Search" field in order to the query being propagated into LDAP.

At least that's how it always worked in the past since the early days of Keycloak.

Reply:
Thank you @mposolda, @mhajas!

Current behavior seems to be following:

- when federated provider is registered within realm and `Users` tab is accessed in admin console there is following message displayed: `This realm may have a federated provider. Viewing all users may cause the system to slow down, but it can be done by searching for "*". Please search for a user above.` and no query is performed. 
- LDAP import mode doesn't seem to have any affect whatsoever. So when searched for `*` users from LDAP are shown regardless of what the setting is
- `Pagination` seems to a flag of the federation provider and it depends on LDAP server itself if the pagination is supported or not
![Screenshot_2023-04-11_18-21-13](https://user-images.githubusercontent.com/1897897/231226655-23f5d8ae-3abe-4618-8c52-ee26d05fbcf7.png)
- when `Pagination` config is set to true there has to be also set some positive limit for the query:
https://github.com/keycloak/keycloak/blob/a08ffdd293b7e4334a92b4caa6b6003fd6a046e0/federation/ldap/src/main/java/org/keycloak/storage/ldap/idm/store/ldap/LDAPIdentityStore.java#L276-L280 When I search for usages of `LDAPQuery.setLimit()` I can see there is possibly non-zero limit set from
  - `LDAPUtils.loadAllLDAPObjects` 
  - `LDAPStorageProviderFactory.syncImpl`
None of the above seem to be used when querying by `searchForUsersStream` method.
- Search in LDAP works only when searched for whole first name, last name, email or username. So e.g. if there is user with username `bwilson` and searched string is `bwils` the user is not found. From JPA such a user it's found and returned.

cc @hmlnarik